### ISSUE TITLE
Docker issue with certain flags

### ISSUE THREAD

**Issue by nicnatros** [ldez/traefik-certs-dumper#179]
I'm using traefik-certs-dumper in a Docker container. When I add certain flags to the entrypoint the app breaks.

The below works.

sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs

When I add the extra flags it breaks.

sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs traefik-certs-dumper file **--crt-ext=.pem --key-ext=.pem –domain-subdir=true**

**Comment by ldez** [comment]
Hello,

your 2nd line contains 2 times "traefik-certs-dumper file".

I don't know if it's a copy-paste error inside the issue or something else.

**Comment by nicnatros** [comment]
Hi Ludovic,
Thank you so much for the response. Yes, that was a copy paste mistake. 

I've corrected the entrypoint to the below and it still doesn't work. Not sure if maybe my syntax is incorrect somewhere.

sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs --crt-ext=.pem --key-ext=.pem –-domain-subdir=true

It runs just fine until I add those last three flags at the end. The problem appears to be there. Can these be specified with environment variables instead? I checked the log but there doesn't seem to be an indication of why it fails. I don't know that it matters but I'm modifying this field in Portainer where I get a status code 400.

fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz

fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz

(1/2) Installing oniguruma (6.9.4-r1)

(2/2) Installing jq (1.6-r0)

Executing busybox-1.30.1-r3.trigger

OK: 7 MiB in 16 packages

**Comment by ldez** [comment]
What is the problem exactly?

**Comment by ldez** [comment]
```console
$ traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs --crt-ext=.pem --key-ext=.pem –-domain-subdir=true
Error: unknown command "–-domain-subdir=true" for "traefik-certs-dumper"
Run 'traefik-certs-dumper --help' for usage.
2022/07/27 20:13:52 root.go:45: unknown command "–-domain-subdir=true" for "traefik-certs-dumper"
```

**Comment by ldez** [comment]
```console
$ traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs --crt-ext=.pem --key-ext=.pem –-domain-subdir=true
Error: --crt-ext (".pem") and --key-ext (".pem") are identical, in this case --domain-subdir is required
Usage:
  traefik-certs-dumper file [flags]

Flags:
  -h, --help             help for file
      --source string    Path to 'acme.json' file. (default "./acme.json")
      --version string   Traefik version. If empty use v1. Possible values: 'v2'.

Global Flags:
      --clean              Clean destination folder before dumping content. (default true)
      --config string      config file (default is $HOME/.traefik-certs-dumper.yaml)
      --crt-ext string     The file extension of the generated certificates. (default ".crt")
      --crt-name string    The file name (without extension) of the generated certificates. (default "certificate")
      --dest string        Path to store the dump content. (default "./dump")
      --domain-subdir      Use domain as sub-directory.
      --key-ext string     The file extension of the generated private keys. (default ".key")
      --key-name string    The file name (without extension) of the generated private keys. (default "privatekey")
      --post-hook string   Execute a command only if changes occurs on the data source. (works only with the watch mode)
      --watch              Enable watching changes.

2022/07/27 20:18:42 root.go:45: --crt-ext (".pem") and --key-ext (".pem") are identical, in this case --domain-subdir is required
exit status 1
```

**Comment by ldez** [comment]
The problem seems related to the flag order, you can use the following command:

```
traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem 
```

**Comment by nicnatros** [comment]
I was using Portainer to do this but that obscured the problem because it's not very verbose. I tried adding the entrypoint using the command line and now there is better feedback on why it isn't running.

root@NAS:/var/lib/docker/bindmounts# docker run --entrypoint sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem traefik-certs-dumper
Unable to find image 'apk:latest' locally
docker: Error response from daemon: pull access denied for apk, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

**Comment by ldez** [comment]
your problem is here `–-domain-subdir=true` you must use this `--domain-subdir=true`.
It's related to a wrong character.

**Comment by nicnatros** [comment]
Good catch, and I did fix the character. Using the proper character there's still an error about the "apk add jq"

Unable to find image 'apk:latest' locally
docker: Error response from daemon: pull access denied for apk, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

root@NAS:/var/lib/docker/bindmounts# docker run --entrypoint sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ jq ".[] | .Certificates | length" /data/acme.json != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem traefik-certs-dumper
Unable to find image 'apk:latest' locally
docker: Error response from daemon: pull access denied for apk, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

**Comment by ldez** [comment]
I think you are not using an alpine image: apk is the package manager for alpine.

**Comment by nicnatros** [comment]
Ok. I didn't realize that this was OS dependent. I'm on Debian.

**Comment by nicnatros** [comment]
Looks like I need to substitute that command for (apt install jq -y) for Debian.

**Comment by ldez** [comment]
or just use an alpine image.

**Comment by ldez** [comment]
you just have to use my image `ldez/traefik-certs-dumper`

**Comment by nicnatros** [comment]
The image I've been using is ldez/traefik-certs-dumper:v2.7.4

docker: Error response from daemon: pull access denied for apt, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

This makes it sound like the container either doesn't have the repository or the container permissions don't allow access to pull a package.

**Comment by ldez** [comment]
No, if you are using the right image, the image contains apk.

The latest version is `v2.8.1` but `v2.7.4` works too.

**Comment by ldez** [comment]
I don't know what you are doing but you are doing something wrong:
```
$ docker run --rm -it --entrypoint /bin/sh ldez/traefik-certs-dumper:v2.8.1
/ # apk -h
apk-tools 2.12.7, compiled for x86_64.

usage: apk [<OPTIONS>...] COMMAND [<ARGUMENTS>...]

Package installation and removal:
add        Add packages to WORLD and commit changes
del        Remove packages from WORLD and commit changes

System maintenance:
fix        Fix, reinstall or upgrade packages without modifying WORLD
update     Update repository indexes
upgrade    Install upgrades available from repositories
cache      Manage the local package cache

Querying package information:
info       Give detailed information about packages or repositories
list       List packages matching a pattern or other criteria
dot        Render dependencies as graphviz graphs
policy     Show repository policy for packages
search     Search for packages by name or description

Repository maintenance:
index      Create repository index file from packages
fetch      Download packages from global repositories to a local directory
manifest   Show checksums of package contents
verify     Verify package integrity and signature

Miscellaneous:
audit      Audit system for changes
stats      Show statistics about repositories and installations
version    Compare package versions or perform tests on version strings

This apk has coffee making abilities.
For more information: man 8 apk
/ # apk add jq
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
(1/2) Installing oniguruma (6.9.7.1-r0)
(2/2) Installing jq (1.6-r1)
Executing busybox-1.34.1-r3.trigger
OK: 7 MiB in 16 packages

```

**Comment by ldez** [comment]
It's because you forget to add the name of the image at the end of the command line.

Try this:
```console
$ docker run --entrypoint='sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem' ldez/traefik-certs-dumper:v2.8.1
```

**Comment by nicnatros** [comment]
That's right. I accidentally inserted container name instead of image name. I think we are very close. 

root@NAS:/var/lib/docker/bindmounts# docker run --entrypoint='sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem' ldez/traefik-certs-dumper:v2.8.1
Unable to find image 'ldez/traefik-certs-dumper:v2.8.1' locally
v2.8.1: Pulling from ldez/traefik-certs-dumper
Digest: sha256:2734462fc8290feab0e34a00d5f3e4992ba74f8e3712d74efff7ae703d0e7e87
Status: Downloaded newer image for ldez/traefik-certs-dumper:v2.8.1
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq \".[] | .Certificates | length\" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem": stat sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem: no such file or directory: unknown.
ERRO[0001] error waiting for container: context canceled

**Comment by ldez** [comment]
you have to mount the directory

**Comment by nicnatros** [comment]
Sorry, this is a result of not being familiar with docker in the command line. I usually use Portainer for the UI to manage containers but wasn't able to get a clear indication of the issue in portainer. In my portainer issue I have the volume bind mounted correctly.

**Comment by ldez** [comment]
If you don't know well Docker, I recommend using a docker-compose file.

You have an example here: https://github.com/ldez/traefik-certs-dumper/blob/master/docs/docker-compose-traefik-v2.yml

**Comment by nicnatros** [comment]
I've had no shortage of dumb mistakes. I believe I might be creating the container right this time, however the error is "no such file or directory: unknown." 

root@NAS:/var/lib/docker/bindmounts/traefik# docker run --name cert-extract4 -v /var/lib/docker/bindmounts/traefik:/data --entrypoint='sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem' ldez/traefik-certs-dumper:v2.8.1
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq \".[] | .Certificates | length\" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem": stat sh,-c, apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs –-domain-subdir=true --crt-ext=.pem --key-ext=.pem: **no such file or directory: unknown.**
ERRO[0000] error waiting for container: context canceled

**Comment by ldez** [comment]
```
docker run --rm  -v /var/lib/docker/bindmounts/traefik/:/data/ --entrypoint='sh' ldez/traefik-certs-dumper:v2.8.1 -c 'apk add jq ; while ! [ -e /data/acme.json ] || ! [ `jq ".[] | .Certificates | length" /data/acme.json` != 0 ]; do sleep 1 ; done && traefik-certs-dumper file --version v2 --watch --source /data/acme.json --dest /data/certs --domain-subdir=true --crt-ext=.pem --key-ext=.pem'
```

**Comment by nicnatros** [comment]
That worked. I can't believe that wrong character re-entered my string. Thank you so much for all your help. I really appreciate it!
